monai/data/dataset.py

```diff
@@ -30,33 +34,116 @@ class Dataset(torch.utils.data.Dataset):
          },                           },                           }]
     """
 
-    def __init__(self, data, transform=None):
+    def __init__(self, data, transform=None, cache_dir=None):
         """
         Args:
             data (Iterable): input data to load and transform to generate dataset for model.
             transform (Callable, optional): transforms to execute operations on input data.
+            cache_dir (Path or str or None): If specified, this is the location for persistent storage
+                of pre-computed transformed data tensors. The cache_dir is computed once, and
+                persists on disk until explicitly removed.  Different runs, programs, experiments
+                may share a common cache dir provided that the trasnsforms pre-processing is
+                consistent.
         """
         self.data = data
+        if not isinstance(transform, Compose):
+            if not isinstance(transform, Iterable):
+                transform = Compose([transform])
+            else:
+                transform = Compose(transform)
         self.transform = transform
+        self.cache_dir = Path(cache_dir) if cache_dir is not None else None
+
+    def _pre_first_random_transform(self, item_transformed):
+        """
+        Process the data from original state up to the first random element.
+
+        Args:
+            item_transformed: The data to be transformed
+        Returns:
+            the transformed element up to the first identified
+            random transform object
+        """
+        for _transform in self.transform.transforms:
+            # execute all the deterministic transforms before the first random transform
+            if isinstance(_transform, Randomizable):
+                break
+            item_transformed = apply_transform(_transform, item_transformed)
+        return item_transformed
+
+    def _first_random_and_beyond_transform(self, item_transformed):
+        """
+        Process the data from before the first random transform to the final state ready for evaluation.
+        Args:
+            item_transformed: The data to be transformed (already process upto the first random transform)
+        Returns:
+            the transformed element through the random transforms
+        """
+        start_post_randomize_run = False
+        for _transform in self.transform.transforms:
+            if start_post_randomize_run or isinstance(_transform, Randomizable):
+                start_post_randomize_run = True
+                item_transformed = apply_transform(_transform, item_transformed)
+        return item_transformed
+
+    def _pre_first_random_transformed_cached_instance(self, item_transformed):
```

Comment:
The function name is a little bit too long.

Reply:
Suggestions?  I'll change it to whatever you want. I was trying to avoid:

```python
# The pfrtci function is the pre first random transformed cached instance
_pfrtci(...)
```

Prefer long descriptive names to short names that require a comment to understand.

Reply:
is `_pre_first_random_transformed` OK?